File: backend/components/metiers/administration_publique/tests/utils/core/core_utils.test.js
// core_utils.test.js – Tests unitaires ultra avancés pour core_utils.js
const core = require('./core_utils');

describe('core_utils – Génération d’ID', () => {
  it('génère un ID unique avec préfixe', () => {
    const id = core.generateId('asset');
    expect(id.startsWith('asset-')).toBe(true);
    expect(id.length).toBeGreaterThan(10);
  });
});

describe('core_utils – deepClone', () => {
  it('clone profondément un objet', () => {
    const obj = { a: 1, b: { c: 2 } };
    const clone = core.deepClone(obj);
    expect(clone).toEqual(obj);
    expect(clone).not.toBe(obj);
    expect(clone.b).not.toBe(obj.b);
  });
});

describe('core_utils – isValidEmail', () => {
  it('valide les emails corrects', () => {
    expect(core.isValidEmail('test@dihya.io')).toBe(true);
  });
  it('rejette les emails invalides', () => {
    expect(core.isValidEmail('not-an-email')).toBe(false);
  });
});

describe('core_utils – auditLog', () => {
  it('génère un log d’audit conforme', () => {
    const log = core.auditLog('LOGIN', 'user-001', { ip: '127.0.0.1' });
    expect(log).toHaveProperty('id');
    expect(log).toHaveProperty('action', 'LOGIN');
    expect(log).toHaveProperty('user', 'user-001');
    expect(log).toHaveProperty('meta');
    expect(log).toHaveProperty('timestamp');
  });
});

File: backend/components/metiers/administration_publique/tests/utils/core/core_utils.test.py
# core_utils.test.py – Tests unitaires ultra avancés pour core_utils.py
import pytest
from .core_utils import generate_id, deep_clone, is_valid_email, audit_log

def test_generate_id():
    id1 = generate_id('asset')
    id2 = generate_id('asset')
    assert id1.startswith('asset-')
    assert id1 != id2

def test_deep_clone():
    obj = {'a': 1, 'b': {'c': 2}}
    clone = deep_clone(obj)
    assert clone == obj
    assert clone is not obj
    assert clone['b'] is not obj['b']

def test_is_valid_email():
    assert is_valid_email('test@dihya.io')
    assert not is_valid_email('not-an-email')

def test_audit_log():
    log = audit_log('LOGIN', 'user-001', {'ip': '127.0.0.1'})
    assert log['action'] == 'LOGIN'
    assert log['user'] == 'user-001'
    assert 'timestamp' in log
    assert log['meta']['ip'] == '127.0.0.1'
